src/Util/Visitor/ReturnNodeVisitor.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace BetterReflection\Util\Visitor;
+
+use PhpParser\NodeVisitorAbstract;
+use PhpParser\Node;
+
+class ReturnNodeVisitor extends NodeVisitorAbstract
+{
+    /**
+     * @var Node\Stmt\Return_[]
+     */
+    private $returnNodes = [];
+
+    private $scopeDepth = 1;
+
+    public function enterNode(Node $node)
+    {
+        if ($this->isScopeChangingNode($node)) {
+            $this->scopeDepth++;
+        }
+
+        if ($this->scopeDepth !== 1) {
+            return;
+        }
+
+        if ($node instanceof Node\Stmt\Return_) {
+            array_push($this->returnNodes, $node);
+        }
+    }
+
+    public function leaveNode(Node $node)
+    {
+        if ($this->isScopeChangingNode($node)) {
+            $this->scopeDepth--;
+        }
+    }
+
+    private function isScopeChangingNode(Node $node)
+    {
+        return $node instanceof Node\Expr\Closure || $node instanceof Node\Stmt\Class_;
```

Comment:
PHP also allows you to nest real (non-closure) functions. A more robust variant would be to check against the FunctionLike interface here.